lib_test/test_ipaddr_b128.ml

```diff
@@ -28,53 +28,72 @@ let assert_raises ~msg exn test_fn =
           Printexc.print_backtrace stderr);
         raise rtexn)
 
-let assert_equal = assert_equal ~printer:Ipaddr_internal.B128.to_string
+let int_of_hex_char c =
+  match c with
+  | '0' .. '9' -> Char.code c - 48
+  | 'a' .. 'f' -> Char.code c - 87
+  | 'A' .. 'F' -> Char.code c - 55
+  | _ -> invalid_arg "char is not a valid hex digit"
+
+let to_string (s : Ipaddr_internal.S128.t) =
+  let s : string = Obj.magic s in
+  List.init 16
+    (fun i -> Printf.sprintf "%.2x" (String.get_uint8 s i))
+  |> String.concat ""
+
+let of_string_exn s : B128.t =
+  if String.length s <> 32 then invalid_arg "not 32 chars long";
+  Bytes.init 16
+    (fun bi ->
+       let i = bi * 2 in
+       let x = int_of_hex_char s.[i+1] and y = int_of_hex_char s.[i] in
+       char_of_int ((y lsl 4) + x))
+  |> Bytes.unsafe_to_string
+  |> Obj.magic
```

Comment:
Here be `Obj.magic`!  but it's only in the tests for types we 'control'. It should be fine. The functions used to live in `B128`, but were only used by the tests.